### PR TITLE
HOCS-5525 - Partial Elastic Search updates

### PR DESCRIPTION
**Pull request by SWhitlam**
This enables correspondents, topics and somu to be partially updated rather than updating the whole case on every update. 

This should resolve issues seen where update messages arrive out of order leading to missing correspondents in Elastic Search.

### CODE REVIEW

File: src/main/java/uk/gov/digital/ho/hocs/search/api/CaseDataService.java [hunk 1]
```diff
@@ -40,6 +40,10 @@ public class CaseDataService {
 
     private final int resultsLimit;
 
+    private final Set<String> correspondentFields = Set.of("allCorrespondents", "currentCorrespondents");
```
Comment (SWhitlam) [inline src/main/java/uk/gov/digital/ho/hocs/search/api/CaseDataService.java:43]: I don't like this but the only alternative is reflection or creating specific methods inside the ES client.

File: src/main/java/uk/gov/digital/ho/hocs/search/client/elasticsearchclient/ElasticSearchClient.java [hunk 2]
```diff
@@ -16,16 +15,9 @@
...
 import uk.gov.digital.ho.hocs.search.domain.model.CaseData;
-
 import java.io.IOException;
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.Map;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
```
Comment (LiamMacP) [inline src/main/java/uk/gov/digital/ho/hocs/search/client/elasticsearchclient/ElasticSearchClient.java:19]: We shouldn't be bringing imports for entire packages? There is an IntelliJ setting that can be changed to stop this.

File: src/main/java/uk/gov/digital/ho/hocs/search/client/elasticsearchclient/ElasticSearchClient.java [hunk 3]
```diff
@@ -49,7 +41,7 @@ public ElasticSearchClient(ObjectMapper objectMapper,
     }
 
     public CaseData findById(UUID uuid) {
-        GetRequest getRequest = new GetRequest(index, uuid.toString());
+       GetRequest getRequest = new GetRequest(index, uuid.toString());
```
Comment (LiamMacP) [inline src/main/java/uk/gov/digital/ho/hocs/search/client/elasticsearchclient/ElasticSearchClient.java:44]: Did you do this to trigger OCD?

File: src/main/java/uk/gov/digital/ho/hocs/search/client/elasticsearchclient/ElasticSearchClient.java [hunk 4]
```diff
@@ -96,6 +88,18 @@ public void update(CaseData caseData) {
         }
     }
 
+    public void update(Set<String> keys, CaseData caseData) {
+        UpdateRequest updateRequest = new UpdateRequest(index, caseData.getCaseUUID().toString());
+        Map<String, Object> documentMapper = objectMapper.convertValue(caseData, Map.class);
```
Comment (LiamMacP) [inline src/main/java/uk/gov/digital/ho/hocs/search/client/elasticsearchclient/ElasticSearchClient.java:93]: I know that we do this in other places, but it does seem like a smell. Is Jackson really not able to handle this automatically somehow?

**Review by LiamMacP — APPROVED** [review]
Just a couple of comments - looks good though!

File: src/main/java/uk/gov/digital/ho/hocs/search/client/elasticsearchclient/ElasticSearchClient.java (hunk 3 above)
Comment (SWhitlam) [reply src/main/java/uk/gov/digital/ho/hocs/search/client/elasticsearchclient/ElasticSearchClient.java:44]: I think IntelliJ did this. We really need a linter!

**Pull request merged**